dev/com.ibm.ws.concurrent.mp_fat/test-applications/MPConcurrentApp/src/concurrent/mp/fat/web/MPConcurrentTestServlet.java

```diff
@@ -4283,6 +4303,156 @@ else if (lookupResult instanceof Throwable)
     }
 
+    /**
+     * Verify exceptionallyAsync(Function) which was added in Java 12
+     */
+    @Test
+    @MinimumJavaLevel(javaLevel = 12)
+    public void testExceptionallyAsync() throws Exception {
+        AtomicInteger count = new AtomicInteger();
+        LinkedBlockingQueue<Object> results = new LinkedBlockingQueue<Object>();
+        String currentThreadName = Thread.currentThread().getName();
+
+        final Function<Throwable, Executor> lookup = (previousFailure) -> {
+            System.out.println("> lookup #" + count.incrementAndGet() + " from testExceptionallyAsync");
+            if (previousFailure != null)
+                results.add(previousFailure);
+            results.add(Thread.currentThread().getName());
+            try {
+                ManagedExecutorService result = InitialContext.doLookup("java:comp/env/executorRef");
+                System.out.println("< lookup: " + result);
+                return result;
+            } catch (NamingException x) {
+                System.out.println("< lookup failed");
+                x.printStackTrace(System.out);
+                throw new CompletionException(x);
+            }
+        };
+
+        String threadName;
+        Object previousFailure;
+
+        // Verify that exceptionally is skipped when no exception is raised by prior stage
+
+        CompletableFuture<?> cf1 = defaultManagedExecutor
+                        .completedFuture((Throwable) null)
+                        .thenApplyAsync(lookup); // expect lookup to succeed because managed executor transfers thread context from the servlet
+        cf1 = exceptionallyAsync.apply(cf1, lookup); // should not be invoked due to lack of any failure in prior stage
+
+        String s;
+        assertTrue(s = cf1.toString(), s.startsWith("ManagedCompletableFuture@"));
+
+        // thenApplyAsync on default execution facility
+        assertNotNull(threadName = results.poll(TIMEOUT_NS, TimeUnit.NANOSECONDS).toString());
+        assertTrue(threadName, threadName.startsWith("Default Executor-thread-")); // must run on Liberty global thread pool
+        assertNotSame(currentThreadName, threadName); // cannot be the servlet thread because operation is async
+
+        assertEquals(defaultManagedExecutor, cf1.get(TIMEOUT_NS, TimeUnit.NANOSECONDS));
+
+        assertEquals(1, count.get()); // lookup function only ran once
+
+        // Verify that exceptionally is invoked when exception is raised by prior stage
+
+        CompletableFuture<?> cf2 = defaultManagedExecutor
+                        .completedFuture((Throwable) null)
+                        .thenApplyAsync(x -> 1 / 0 < 2 ? sameThreadExecutor : null, testThreads); // force ArithmeticException to occur
+        cf2 = exceptionallyAsync.apply(cf2, lookup);
+
+        assertTrue(s = cf2.toString(), s.startsWith("ManagedCompletableFuture@"));
+
+        // exceptionally on unmanaged thread or servlet thread
```

Comment:
This comment doesn't seem true.  Because exceptionallyAsync is used rather than exceptionally, it should neither run on an unmanaged thread nor should it run on the servlet thread.   It looks like test code later on no-ops this check and only checks that the threadName is non-null.  But we could have checked that the thread name doesn't match the servlet thread name and does indicate a managed thread.